lib/Containers/MerkleTree.cpp

```diff
@@ -474,16 +461,67 @@ std::unique_ptr<MerkleTree<BranchingBits, LockStripes>> MerkleTree<BranchingBits
       new MerkleTree<BranchingBits, LockStripes>(*this));
 }
 
+template <std::size_t const BranchingBits, std::size_t const LockStripes>
+std::unique_ptr<MerkleTree<BranchingBits, LockStripes>>
+MerkleTree<BranchingBits, LockStripes>::cloneWithDepth(std::size_t newDepth) {
+  std::shared_lock<std::shared_mutex> guard(_bufferLock);
+
+  if (newDepth == meta().maxDepth) {
+    return std::unique_ptr<MerkleTree<BranchingBits, LockStripes>>(
+        new MerkleTree<BranchingBits, LockStripes>(*this));
+  }
+
+  if (newDepth < meta().maxDepth) {
+    std::unique_ptr<MerkleTree<BranchingBits, LockStripes>> newTree =
+        std::make_unique<MerkleTree<BranchingBits, LockStripes>>(newDepth,
+                                                                 meta().rangeMin,
+                                                                 meta().rangeMax);
+    for (std::size_t index = 0; index < nodeCountUpToDepth(newDepth); ++index) {
+      Node& n = this->node(index);
+      Node& m = newTree->node(index);
+      m = n;
+    }
+    return newTree;
+  }
+
+  // Otherwise let's grow the tree deeper. We're going to grow one level at a
+  // time, recursively. Typically we'll only be requesting to grow one level
+  // deeper at a time anyway, and we should very, very rarely be growing more
+  // than a couple levels at a time.
+  std::uint64_t newRangeMax =
+      meta().rangeMin + ((meta().rangeMax - meta().rangeMin) * BranchingFactor);
+  std::unique_ptr<MerkleTree<BranchingBits, LockStripes>> newTree =
```

Comment:
```suggestion
  auto newTree =
```

Reply:
To be honest, I'd like to avoid the use of `auto`.

Reply:
Feel free to do that. But I'd like to point to this post: https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/
There are some good (also some less convincing) arguments for `auto` there.

Reply:
@goedderz To clarify, I'm not against `auto` in general, just in this file. The type specificity for e.g. `size_t` vs `uint64_t` matters here, and has been a problem with `auto` in the past. Obviously the pointers are different, but it makes it a lot easier to double-check if I can just grep for `auto`.